File: Chap5/C/exo7.c
#include<stdio.h>
#include<pthread.h>
#include<stdlib.h>
#include<time.h>
#include<semaphore.h>
#include<unistd.h>

int compteur=0;
double u;
sem_t sem1;
sem_t sem2;

void *pro1(void* arg)
{
    while(compteur<10)
    {
        sem_wait(&sem1);
        compteur++;
        sem_post(&sem1);
        sem_wait(&sem2);
        u=(1.0/4.0)*((u-1)*(u-1));
        printf("la valeur u dans T1  :%f",u);
        sem_post(&sem2);
        sleep(rand()%5+1);
    }
}
void *pro2(void *arg)
{
    while(compteur<10) //juste pour voir le resulat
    {
        sem_wait(&sem1);
        compteur++;
        sem_post(&sem1);
        sem_wait(&sem2);
        u=(1.0/6.0)*((u-2)*(u-2));
         printf("la valeur de u dans T2 :%f",u);
        sem_post(&sem2);
        sleep(rand()%5+1);
    }    
}

int main()
{
    pthread_t thread1,thread2;
    u=1;
    srand(time(NULL));
    sem_init(&sem1, 0, 1);
    sem_init(&sem2, 0, 1);
    pthread_create(&thread1,NULL,pro1,NULL);
    pthread_create(&thread2,NULL,pro2,NULL);
    pthread_join(thread1,NULL);
    pthread_join(thread2,NULL);
    sem_destroy(&sem1);
    sem_destroy(&sem2);
}

File: Chap5/Python/exo7.py
import threading
import random
import time

compteur = 0
u = 1.0

sem1 = threading.Semaphore(1)
sem2 = threading.Semaphore(1)

def pro1():
    global compteur, u
    while compteur < 10:
        sem1.acquire()
        compteur += 1
        sem1.release()

        sem2.acquire()
        u = (1.0 / 4.0) * ((u - 1) * (u - 1))
        print(f"la valeur u dans T1 : {u}")
        sem2.release()

        time.sleep(random.randint(1, 5))

def pro2():
    global compteur, u
    while compteur < 10:
        sem1.acquire()
        compteur += 1
        sem1.release()

        sem2.acquire()
        u = (1.0 / 6.0) * ((u - 2) * (u - 2))
        print(f"la valeur de u dans T2 : {u}")
        sem2.release()

        time.sleep(random.randint(1, 5))

    global u
    random.seed()
    u = 1.0

    thread1 = threading.Thread(target=pro1)
    thread2 = threading.Thread(target=pro2)

    thread1.start()
    thread2.start()

    thread1.join()
    thread2.join()
